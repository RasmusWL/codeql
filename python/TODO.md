We have gotten quite a lot of QL classes now... and their names doesn't map very well to the call types they handle.

left some TODOs in the code right now.

First item on the agenda is to, in both (datamodel, calls/test.py, and the old callgraph testing), add empty subclass and see that calls do not get resolved... obviously the current handling of `C.foo(C, arg0)` is bad, since it's not class based, but function based... and that will not handle subclass shit...

Can definitly do a grouping that is (1) plain functions (2) method calls on classes (bound-method, classmethod, staticmethod, direct call of method (non-bound))
