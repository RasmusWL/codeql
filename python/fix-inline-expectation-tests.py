#!/usr/bin/env python3

import sys
from pathlib import Path
from dataclasses import MISSING, dataclass
from collections import defaultdict
from enum import Enum
from typing import List
import re

EXPECTATION_RE = r"#\s*\$.*$"

@dataclass
class FileSpec:
    filename: str
    start_line: int
    start_column: int
    end_line: int
    end_colum: int

    def __init__(self, filename, start_line, start_column, end_line, end_colum):
        self.filename = filename.strip()
        self.start_line = int(start_line)
        self.start_column = int(start_column)
        self.end_line = int(end_line)
        self.end_colum = int(end_colum)

class ProblemType(Enum):
    FIXED_MISSING = "Fixed missing result:"
    UNEXPECTED_RESULT = "Unexpected result: "
    MISSING_RESULT = "Missing result:"

class HandleMissing(Enum):
    REMOVE = "remove"
    ADD_TO_MISSING = "add"

@dataclass
class Problem:
    problem: ProblemType
    tag: str
    value: str

    def format_expectation(self) -> str:
        if self.value == "":
            return self.tag
        else:
            return f"{self.tag}={self.value}"

    @classmethod
    def from_input(cls, raw: str):
        raw = raw.strip()
        for pt in ProblemType:
            if raw.startswith(pt.value):
                break
        else:
            raise ValueError("Not a known problem type", raw)

        tag, value = raw[len(pt.value):].split("=")

        return cls(pt, tag, value)


def fix_concept_tests(ct_expected_path: Path, handle_missing: HandleMissing = HandleMissing.REMOVE):

    problems_in_file = defaultdict(lambda: defaultdict(list))

    lines = ct_expected_path.read_text().splitlines()

    if lines[0].startswith("|"):
        inline_expectation_lines = lines
    else:
        # there are multiple query predicates
        start = None
        end = None

        for i, line in enumerate(lines):
            if start is None:
                if line == "failures":
                    start = i + 1
            else:
                if not line.startswith("|"):
                    end = i
                    break

        assert start is not None
        inline_expectation_lines = lines[start:end]

    for line in inline_expectation_lines:
        _, file_spec_raw, _, problem_raw, _ = line.split("|")

        spec = FileSpec(*file_spec_raw.split(":"))

        if spec.start_line != spec.end_line:
            print(f"skipping line {spec.start_line}-{spec.end_line} of {spec.filename} since this script doesn't have multiline support: {line}", file=sys.stderr)
            continue

        problem = Problem.from_input(problem_raw)

        problems_in_file[spec.filename][spec.start_line].append(problem)

    for filename, problems_on_line in problems_in_file.items():
        with ct_expected_path.with_name(filename).open('r+') as f:
            lines = f.readlines()

            for line_num, problems in problems_on_line.items():
                line_index = line_num-1
                problems: List[Problem]
                stuff_we_can_handle = [ProblemType.UNEXPECTED_RESULT]
                if not all(p.problem in stuff_we_can_handle for p in problems):
                    print(f"Skipping line {line_num} of {filename} since not all problems are {stuff_we_can_handle}", file=sys.stderr)
                    continue

                line = lines[line_index]
                if (match := re.search(EXPECTATION_RE, line)):
                    new_expectations = [match.group(0)]
                    line = line[:match.start()]
                else:
                    new_expectations = [" # $"]
                    line = line.rstrip()

                for problem in problems:
                    if problem.problem == ProblemType.UNEXPECTED_RESULT:
                        to_add = problem.format_expectation()
                        if to_add not in new_expectations:
                            new_expectations.append(to_add)
                    else:
                        raise Exception("problem type not handled", problem)

                new_expecatations_formatted = " ".join(new_expectations)
                lines[line_index] = line + new_expecatations_formatted

            f.truncate()
            f.seek(0)
            for line in lines:
                f.write(line.rstrip("\n"))
                f.write("\n")


if __name__ == "__main__":
    ct_in = Path(sys.argv[1])
    fix_concept_tests(ct_in)
