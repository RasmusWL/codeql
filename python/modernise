#!/bin/bash

set -e

CURRENTDIR=$(pwd)
SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
PATTERNS="$SCRIPTDIR/modernise-patterns"

name="$1"
test=$(readlink -f "$2")
files="${@:3}"
auto_branch_name="python-modernise-$(echo $name | tr ' ' '-')"

if [ -z $SEMMLE_CODE ]; then
    echo "\$SEMMLE_CODE must be set to your 'code' repo"
    exit 1
fi

# TODO: Understand the case syntax :D

function should_run {
    echo -n "$@? "
    read -p "[Yn]" yn
    case $yn in
        [Nn]* ) /bin/false;;
        * ) /bin/true;;
    esac
}

function should_run_false {
    echo -n "$@? "
    read -p "[yN]" yn
    case $yn in
        [Yy]* ) /bin/true;;
        * ) /bin/false;;
    esac
}

for file in $files; do
    echo $file
done


if should_run "new branch (${auto_branch_name})"; then
    git checkout -b ${auto_branch_name}
fi

if should_run "auto format"; then
    pyqlformat $files

    if should_run "commit auto format"; then
    git commit -am "Python: Autoformat ${name}"
    fi
fi


if should_run "modernise (regexp replace)"; then
    repren -p "$PATTERNS" $files
    repren -p "$PATTERNS" $files
    repren -p "$PATTERNS" $files

    for file in $files; do
        rm -f $file.orig
    done
fi

if should_run "run tests on ${test}"; then
    if [[ -d "${test}" ]]; then
        test_folder="${test}"
    else
        test_folder=$(dirname "${test}")
    fi

    while :; do
        cd $SEMMLE_CODE && ./target/jars/qltest --threads 8 "${test}" || ~/bin/qlicdiff "${test_folder}"
        if should_run_false "accept results"; then
            rename -f -e 's/\.actual$/.expected/' ${test_folder}/*.actual
            break
        fi
    done
fi

cd "${CURRENTDIR}"

if should_run "commit modernise"; then
    git commit -am "Python: Modernise ${name}"
fi

if should_run "make pull-request"; then
    current_branch_name=$(git rev-parse --abbrev-ref HEAD)
    git push origin ${current_branch_name}
    xdg-open "https://github.com/RasmusWL/ql/pull/new/${current_branch_name}"
    git checkout master
fi
