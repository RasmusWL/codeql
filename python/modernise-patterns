\(FunctionObject\)\.getFunction\(\)	(FunctionValue).getScope()
\.getPyClass\(\)	.getScope()

# -> Value::named
ModuleObject::named\("(.*)"\).attr\("(.*)"\)	Value::named("\1.\2")
ModuleObject::named\("(.*)"\)	Value::named("\1")
Module::named\("(.*)"\).attr\("(.*)"\)	Value::named("\1.\2")

# Object -> Value
\bModuleObject\b	ModuleValue
\bFunctionObject\b	FunctionValue
\bClassObject\b	ClassValue
\bStringObject\b	StringValue
\bNumericObject\b	NumericValue
\bTupleObject\b	TupleValue
\bObject\b	Value

\bPyFunctionObject\b	PythonFunctionValue

\.lookupAttribute\(	.lookup(

# refersTo -> pointsTo
\.refersTo\(_, (.*), _\)	.pointsTo().getClass() = \1
\.refersTo\((.*)\)	.pointsTo(\1)

\.getASuperType\(\)	.getABaseType+()
