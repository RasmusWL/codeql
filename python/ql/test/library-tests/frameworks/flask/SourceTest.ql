/**
 * @kind path-problem
 */

import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.dataflow.new.TaintTracking
import semmle.python.dataflow.new.RemoteFlowSources
import DataFlow::PathGraph

// see `source_test.py`
class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "SourceTest" }

  override predicate isSource(DataFlow::Node source) {
    source instanceof RemoteFlowSource and
    source.getLocation().getFile().getShortName().matches("source_test%")
  }

  override predicate isSink(DataFlow::Node sink) {
    exists(CallNode call |
      call.getFunction().(NameNode).getId() = "SINK" and
      sink.(DataFlow::CfgNode).getNode() = call.getAnArg() and
      sink.getLocation().getFile().getShortName().matches("source_test%")
    )
  }
}

from Configuration config, DataFlow::PathNode source, DataFlow::PathNode sink
where config.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "Path text"
