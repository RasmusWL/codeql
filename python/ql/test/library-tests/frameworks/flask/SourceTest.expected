edges
| source_test2.py:16:17:16:23 | ControlFlowNode for request | source_test2.py:17:27:17:33 | ControlFlowNode for request |
| source_test2.py:16:17:16:23 | ControlFlowNode for request | source_test2.py:17:27:17:38 | ControlFlowNode for Attribute |
| source_test2.py:17:27:17:33 | ControlFlowNode for request | source_test2.py:17:27:17:38 | ControlFlowNode for Attribute |
| source_test2.py:17:27:17:38 | ControlFlowNode for Attribute | source_test2.py:17:27:17:45 | ControlFlowNode for Subscript |
| source_test2.py:17:27:17:45 | ControlFlowNode for Subscript | source_test2.py:18:14:18:28 | ControlFlowNode for user_controlled |
| source_test2.py:24:27:24:39 | ControlFlowNode for Attribute | source_test2.py:24:27:24:44 | ControlFlowNode for Attribute |
| source_test2.py:24:27:24:44 | ControlFlowNode for Attribute | source_test2.py:24:27:24:51 | ControlFlowNode for Subscript |
| source_test2.py:24:27:24:51 | ControlFlowNode for Subscript | source_test2.py:25:14:25:28 | ControlFlowNode for user_controlled |
| source_test.py:9:17:9:23 | ControlFlowNode for request | source_test.py:10:27:10:33 | ControlFlowNode for request |
| source_test.py:9:17:9:23 | ControlFlowNode for request | source_test.py:10:27:10:38 | ControlFlowNode for Attribute |
| source_test.py:10:27:10:33 | ControlFlowNode for request | source_test.py:10:27:10:38 | ControlFlowNode for Attribute |
| source_test.py:10:27:10:38 | ControlFlowNode for Attribute | source_test.py:10:27:10:45 | ControlFlowNode for Subscript |
| source_test.py:10:27:10:45 | ControlFlowNode for Subscript | source_test.py:11:14:11:28 | ControlFlowNode for user_controlled |
| source_test.py:17:27:17:33 | ControlFlowNode for request | source_test.py:17:27:17:38 | ControlFlowNode for Attribute |
| source_test.py:17:27:17:38 | ControlFlowNode for Attribute | source_test.py:17:27:17:45 | ControlFlowNode for Subscript |
| source_test.py:17:27:17:45 | ControlFlowNode for Subscript | source_test.py:20:10:20:24 | ControlFlowNode for user_controlled |
| source_test.py:19:27:19:33 | ControlFlowNode for request | source_test.py:19:27:19:38 | ControlFlowNode for Attribute |
| source_test.py:19:27:19:38 | ControlFlowNode for Attribute | source_test.py:19:27:19:45 | ControlFlowNode for Subscript |
| source_test.py:19:27:19:45 | ControlFlowNode for Subscript | source_test.py:20:10:20:24 | ControlFlowNode for user_controlled |
nodes
| source_test2.py:16:17:16:23 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| source_test2.py:17:27:17:33 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| source_test2.py:17:27:17:38 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| source_test2.py:17:27:17:45 | ControlFlowNode for Subscript | semmle.label | ControlFlowNode for Subscript |
| source_test2.py:18:14:18:28 | ControlFlowNode for user_controlled | semmle.label | ControlFlowNode for user_controlled |
| source_test2.py:24:27:24:39 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| source_test2.py:24:27:24:44 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| source_test2.py:24:27:24:51 | ControlFlowNode for Subscript | semmle.label | ControlFlowNode for Subscript |
| source_test2.py:25:14:25:28 | ControlFlowNode for user_controlled | semmle.label | ControlFlowNode for user_controlled |
| source_test.py:9:17:9:23 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| source_test.py:10:27:10:33 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| source_test.py:10:27:10:38 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| source_test.py:10:27:10:45 | ControlFlowNode for Subscript | semmle.label | ControlFlowNode for Subscript |
| source_test.py:11:14:11:28 | ControlFlowNode for user_controlled | semmle.label | ControlFlowNode for user_controlled |
| source_test.py:17:27:17:33 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| source_test.py:17:27:17:38 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| source_test.py:17:27:17:45 | ControlFlowNode for Subscript | semmle.label | ControlFlowNode for Subscript |
| source_test.py:19:27:19:33 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| source_test.py:19:27:19:38 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| source_test.py:19:27:19:45 | ControlFlowNode for Subscript | semmle.label | ControlFlowNode for Subscript |
| source_test.py:20:10:20:24 | ControlFlowNode for user_controlled | semmle.label | ControlFlowNode for user_controlled |
subpaths
#select
| source_test2.py:18:14:18:28 | ControlFlowNode for user_controlled | source_test2.py:16:17:16:23 | ControlFlowNode for request | source_test2.py:18:14:18:28 | ControlFlowNode for user_controlled | Path text |
| source_test2.py:18:14:18:28 | ControlFlowNode for user_controlled | source_test2.py:17:27:17:33 | ControlFlowNode for request | source_test2.py:18:14:18:28 | ControlFlowNode for user_controlled | Path text |
| source_test2.py:25:14:25:28 | ControlFlowNode for user_controlled | source_test2.py:24:27:24:39 | ControlFlowNode for Attribute | source_test2.py:25:14:25:28 | ControlFlowNode for user_controlled | Path text |
| source_test.py:11:14:11:28 | ControlFlowNode for user_controlled | source_test.py:9:17:9:23 | ControlFlowNode for request | source_test.py:11:14:11:28 | ControlFlowNode for user_controlled | Path text |
| source_test.py:11:14:11:28 | ControlFlowNode for user_controlled | source_test.py:10:27:10:33 | ControlFlowNode for request | source_test.py:11:14:11:28 | ControlFlowNode for user_controlled | Path text |
| source_test.py:20:10:20:24 | ControlFlowNode for user_controlled | source_test.py:17:27:17:33 | ControlFlowNode for request | source_test.py:20:10:20:24 | ControlFlowNode for user_controlled | Path text |
| source_test.py:20:10:20:24 | ControlFlowNode for user_controlled | source_test.py:19:27:19:33 | ControlFlowNode for request | source_test.py:20:10:20:24 | ControlFlowNode for user_controlled | Path text |
